braindecode/preprocessing/windowers.py

```diff
@@ -287,6 +292,111 @@ def _create_windows_from_events(
+
+    Parameters
+    ----------
+    ds : BaseDataset
+        Dataset containing continuous data and description.
+    infer_mapping : bool
+        If True, extract all events from all datasets and map them to
+        increasing integers starting from 0.
+    infer_window_size_stride : bool
+        If True, infer the stride from the original trial size of the first
+        trial and trial_start_offset_samples and trial_stop_offset_samples.
+
+    See `create_windows_from_events` for description of other parameters.
+
+    Returns
+    -------
+    WindowsDataset :
+        Windowed dataset.
+    """
+    # catch window_kwargs to store to dataset
+    window_kwargs = [
+        (create_windows_from_events.__name__, _get_windowing_kwargs(locals())),
+    ]
+    if infer_mapping:
+        unique_events = np.unique(ds.raw.annotations.description)
+        new_unique_events = [x for x in unique_events if x not in mapping]
+        # mapping event descriptions to integers from 0 on
+        max_id_mapping = len(mapping)
+        mapping.update(
+            {v: k + max_id_mapping for k, v in enumerate(new_unique_events)}
+        )
+
+    events, events_id = mne.events_from_annotations(ds.raw, mapping)
+    onsets = events[:, 0]
+    # Onsets are relative to the beginning of the recording
+    filtered_durations = np.array(
+        [a['duration'] for a in ds.raw.annotations
+            if a['description'] in events_id]
+    )
+    stops = onsets + (filtered_durations * ds.raw.info['sfreq']).astype(int)
+    # XXX This could probably be simplified by using chunk_duration in
+    #     `events_from_annotations`
+
+    last_samp = ds.raw.first_samp + ds.raw.n_times
+    if stops[-1] + trial_stop_offset_samples > last_samp:
+        raise ValueError(
+            '"trial_stop_offset_samples" too large. Stop of last trial '
+            f'({stops[-1]}) + "trial_stop_offset_samples" '
+            f'({trial_stop_offset_samples}) must be smaller than length of'
+            f' recording ({len(ds)}).')
+
+    if infer_window_size_stride:
+        # window size is trial size
+        if window_size_samples is None:
+            window_size_samples = stops[0] + trial_stop_offset_samples - (
+                onsets[0] + trial_start_offset_samples)
+            window_stride_samples = window_size_samples
+        this_trial_sizes = (stops + trial_stop_offset_samples) - (
+            onsets + trial_start_offset_samples)
+        # Maybe actually this is not necessary?
+        # We could also just say we just assume window size=trial size
+        # in case not given, without this condition...
+        # but then would have to change functions overall
+        # to deal with varying window sizes hmmhmh
+        assert np.all(this_trial_sizes == window_size_samples), (
+            'All trial sizes should be the same if you do not supply a window '
+            'size.')
+
+    description = events[:, -1]
+
```

Comment:
I woudl avoid code duplication here as the code is long. You could either make the first part of the function a sub function, or simply add a parameter to switch from epochs to raws and use `functools.partial(_create_windows_from_events, from_epochs=True))` as the `create_window_fn`.

Reply:
Good idea, done